src/guardian/unvetting/unvetting.service.ts

```diff
@@ -0,0 +1,214 @@
+    private securityService: SecurityService,
+    private guardianMessageService: GuardianMessageService,
+  ) {}
+
+  /**
+   * Handles unvetting of invalid, duplicated, and front-ran keys.
+   *
+   * 1. Collects keys flagged for unvetting from `stakingModuleData`.
+   * 2. Logs and exits if no keys require unvetting.
+   * 3. Retrieves the maximum operators per unvetting from `SecurityService`.
+   * 4. Prepares and processes the first chunk to avoid transaction races.
+   * 5. Sends a transaction to the Security contract and forwards messages to the guardian broker.
+   *
+   * @param stakingModuleData - Staking module data, including keys for unvetting.
+   * @param blockData - Collected data from the current block.
+   * @returns void
+   */
+  public async handleUnvetting(
+    stakingModuleData: StakingModuleData,
+    blockData: BlockData,
+  ) {
+    const invalidKeys = this.collectInvalidKeys(stakingModuleData);
+
+    if (!invalidKeys.length) {
+      this.logNoUnvettingNeeded(blockData, stakingModuleData);
+      return;
+    }
+
+    const maxOperatorsPerUnvetting = await this.getMaxOperatorsPerUnvetting();
+    const firstChunk = this.calculateNewStakingLimit(
+      invalidKeys,
+      maxOperatorsPerUnvetting,
+    );
+
+    await this.processUnvetting(stakingModuleData, blockData, firstChunk);
+  }
+
+  private collectInvalidKeys(
+    stakingModuleData: StakingModuleData,
+  ): RegistryKey[] {
+    return stakingModuleData.invalidKeys.concat(
+      stakingModuleData.duplicatedKeys,
+      stakingModuleData.frontRunKeys,
+    );
+  }
+
+  private logNoUnvettingNeeded(
+    blockData: BlockData,
+    stakingModuleData: StakingModuleData,
+  ): void {
+    this.logger.debug?.('Keys are correct. No need for unvetting', {
+      blockHash: blockData.blockHash,
+      stakingModuleId: stakingModuleData.stakingModuleId,
+    });
+  }
+
+  public async processUnvetting(
+    stakingModuleData: StakingModuleData,
+    blockData: BlockData,
+    chunk: UnvetData,
+  ) {
+    const { blockNumber, blockHash, guardianAddress, guardianIndex } =
+      blockData;
+    const { nonce, stakingModuleId } = stakingModuleData;
+    const { operatorIds, vettedKeysByOperator } = chunk;
+    const signature = await this.securityService.signUnvetData(
+      nonce,
+      blockNumber,
+      blockHash,
+      stakingModuleId,
+      operatorIds,
+      vettedKeysByOperator,
+    );
+
+    if (!blockData.walletBalanceCritical) {
+      this.logSufficientBalance(blockData, stakingModuleData);
+
+      this.securityService
+        .unvetSigningKeys(
+          nonce,
+          blockNumber,
+          blockHash,
+          stakingModuleId,
+          operatorIds,
+          vettedKeysByOperator,
+          signature,
+        )
+        .catch((error) =>
+          this.logger.error('Failed to send unvet transaction', {
+            error,
+            blockHash,
+            stakingModuleId,
+          }),
+        );
```

Comment:
So, all guardians are potentially sending the same tx at the same time?

Reply:
It will be reverted for other councils if one has already been successfully sent